### ISSUE TITLE
Akka HTTP WebSocket connections reset

### ISSUE THREAD

**Issue by jroper** [akka/akka#19467]
In our tests, I'm frequently seeing Akka HTTP resetting TCP connections (that is, sending the RST flag mid connection, not sending a FIN flag).

Here's a typical connection, Akka HTTP is on port 19001, the client sends an HTTP request which is a WebSocket request, Akka HTTP sends a response back accepting it, and then Akka HTTP immediately resets the connection (notice the R flag):

```
13:26:42.273419 IP localhost.34368 > localhost.19001: Flags [S], seq 2355686127, win 43690, options [mss 65495,sackOK,TS val 25598901 ecr 0,nop,wscale 7], length 0
E..<qZ@.@.._.........@J9.h...........0.........
............
13:26:42.273439 IP localhost.19001 > localhost.34368: Flags [S.], seq 2135811301, ack 2355686128, win 43690, options [mss 65495,sackOK,TS val 25598901 ecr 25598901,nop,wscale 7], length 0
E..<..@.@.<.........J9.@.M...h.......0.........
............
13:26:42.273452 IP localhost.34368 > localhost.19001: Flags [.], ack 1, win 342, options [nop,nop,TS val 25598901 ecr 25598901], length 0
E..4q[@.@..f.........@J9.h...M.....V.(.....
........
13:26:42.285304 IP localhost.34368 > localhost.19001: Flags [P.], seq 1:207, ack 1, win 342, options [nop,nop,TS val 25598904 ecr 25598901], length 206
E...q\@.@............@J9.h...M.....V.......
........GET /stream HTTP/1.1
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Key: Rs0VY7XccMnGDIOrFTPykg==
Host: localhost:19001
Sec-WebSocket-Origin: http://localhost:19001
Sec-WebSocket-Version: 13

13:26:42.285320 IP localhost.19001 > localhost.34368: Flags [.], ack 207, win 350, options [nop,nop,TS val 25598904 ecr 25598904], length 0
E..4.S@.@..n........J9.@.M...h.....^.(.....
........
13:26:42.344631 IP localhost.19001 > localhost.34368: Flags [P.], seq 1:192, ack 207, win 350, options [nop,nop,TS val 25598919 ecr 25598904], length 191
E....T@.@...........J9.@.M...h.....^.......
........HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Sec-WebSocket-Accept: mXpkJMRRA9Zm8qiRzzBNd0Yo6wI=
Server: akka-http/2.4.1
Date: Fri, 15 Jan 2016 02:26:42 GMT
Connection: upgrade

13:26:42.344647 IP localhost.34368 > localhost.19001: Flags [.], ack 192, win 350, options [nop,nop,TS val 25598919 ecr 25598919], length 0
E..4q]@.@..d.........@J9.h...M.....^.(.....
........
13:26:42.348685 IP localhost.19001 > localhost.34368: Flags [R.], seq 192, ack 207, win 350, options [nop,nop,TS val 25598920 ecr 25598919], length 0
E..4.U@.@..l........J9.@.M...h.....^.(.....
........
```

I don't know why or how Akka HTTP would or could reset a connection.  On the server side the flow handling this is simply ignoring WebSocket messages, and using a `Source.maybe` to send nothing.  The WebSocket flow is started, but not finished.  The result of this is that the client gets a "Connection reset by peer" exception.  According to [here](https://docs.oracle.com/javase/8/docs/technotes/guides/net/articles/connection_release.html), Akka HTTP must be calling `Socket.setLinger (true, 0)` before closing the connection in order to trigger the TCP reset.

**Comment by jroper** [comment]
I've found out that Akka is definitely causing the RST to be sent, but as for why, I have no idea, I have turned on debug logging and TCP trace logging, and this is all I get:

```
DEBUG a.i.TcpListener - New connection accepted
DEBUG a.i.SelectionHandler - Executing [WorkerForCommand(RegisterIncoming(java.nio.channels.SocketChannel[connected local=/127.0.0.1:19001 remote=/127.0.0.1:35050]),Actor[akka://application/system/IO-TCP/selectors/$a/0#1348593095],<function1>)]
DEBUG a.i.TcpIncomingConnection - [Actor[akka://application/user/StreamSupervisor-4/StageActorRef-5]] registered as connection handler
DEBUG a.i.TcpIncomingConnection - Read [206] bytes.
DEBUG a.i.TcpIncomingConnection - Wrote [191] bytes to channel
Fri Jan 15 13:52:56 AEDT 2016 ENTER websocket-downstream.preStart in application-akka.actor.default-dispatcher-5: 
Fri Jan 15 13:52:56 AEDT 2016 LEAVE websocket-downstream.preStart at -299
Fri Jan 15 13:52:56 AEDT 2016 ENTER websocket-upstream.preStart in application-akka.actor.default-dispatcher-5: 
Fri Jan 15 13:52:56 AEDT 2016 LEAVE websocket-upstream.preStart at -265
DEBUG a.i.TcpIncomingConnection - Got Abort command. RESETing connection.
```

Note, the `ENTER/LEAVE` log messages are being output by a `PushPullStage` that I've injected into the flow passed to `UpgradeToWebsocketLowLevel.handleFrames`, there's one on each side of the flow.  As you can see, nothing, no errors, no nothing is going in our out of that flow.  For some reason, it seems Akka HTTP is deciding of its own accord to abort this connection, and is not logging anything about it.

**Comment by jroper** [comment]
Ok, so I found out where Akka streams is sending the `Abort` message, set a break point in there, and now I can see the bug in Akka HTTP:

```
java.lang.IllegalArgumentException: requirement failed: Cannot pull closed port
	at scala.Predef$.require(Predef.scala:219)
	at akka.stream.stage.GraphStageLogic.pull(GraphStage.scala:419)
	at akka.http.impl.engine.server.HttpServerBluePrint$ProtocolSwitchStage$$anon$4$$anon$19.onPull(HttpServerBluePrint.scala:456)
	at akka.stream.stage.GraphStageLogic$SubSourceOutlet$$anonfun$4.apply(GraphStage.scala:1096)
	at akka.stream.stage.GraphStageLogic$SubSourceOutlet$$anonfun$4.apply(GraphStage.scala:1092)
	at akka.stream.impl.fusing.GraphInterpreter.runAsyncInput(GraphInterpreter.scala:557)
	at akka.stream.impl.fusing.GraphInterpreterShell.receive(ActorGraphInterpreter.scala:399)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:549)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:480)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:495)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
	at akka.actor.ActorCell.invoke(ActorCell.scala:495)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by jroper** [comment]
I've raised a bug about the lack of error reporting here: #19469

**Comment by jroper** [comment]
Let me know if you need more help reproducing, but as far as I can tell, the Akka HTTP server code I'm using is not impacting this, this exception happens before any signals are sent to my WebSocket flows.  I did try to implement a simple reproducer, but was unable to reproduce it.  The tcpdump shows exactly what the client is doing.

**Comment by jroper** [comment]
By the way, the specific stack trace above was produced by this pcap:

```
15:04:04.057765 IP localhost.35820 > localhost.19001: Flags [S], seq 1991301263, win 43690, options [mss 65495,sackOK,TS val 27059347 ecr 0,nop,wscale 7], length 0
E..<<.@.@.............J9v............0.........
............
15:04:04.057784 IP localhost.19001 > localhost.35820: Flags [S.], seq 3436020017, ack 1991301264, win 43690, options [mss 65495,sackOK,TS val 27059347 ecr 27059347,nop,wscale 7], length 0
E..<..@.@.<.........J9.....1v........0.........
............
15:04:04.057796 IP localhost.35820 > localhost.19001: Flags [.], ack 1, win 342, options [nop,nop,TS val 27059347 ecr 27059347], length 0
E..4<.@.@.............J9v......2...V.(.....
........
15:04:04.059397 IP localhost.35820 > localhost.19001: Flags [P.], seq 1:207, ack 1, win 342, options [nop,nop,TS val 27059347 ecr 27059347], length 206
E...<.@.@.............J9v......2...V.......
........GET /stream HTTP/1.1
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Key: WPUQjV+v9cJoL53hARhkmA==
Host: localhost:19001
Sec-WebSocket-Origin: http://localhost:19001
Sec-WebSocket-Version: 13


15:04:04.059411 IP localhost.19001 > localhost.35820: Flags [.], ack 207, win 350, options [nop,nop,TS val 27059347 ecr 27059347], length 0
E..4e.@.@.. ........J9.....2v..^...^.(.....
........
15:04:04.146257 IP localhost.19001 > localhost.35820: Flags [P.], seq 1:192, ack 207, win 350, options [nop,nop,TS val 27059369 ecr 27059347], length 191
E...e.@.@..`........J9.....2v..^...^.......
........HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Sec-WebSocket-Accept: /aAdv2eJm+e2geTnHa0FLcCV+kE=
Server: akka-http/2.4.1
Date: Fri, 15 Jan 2016 04:04:04 GMT
Connection: upgrade


15:04:04.146275 IP localhost.35820 > localhost.19001: Flags [.], ack 192, win 350, options [nop,nop,TS val 27059369 ecr 27059369], length 0
E..4<.@.@.............J9v..^.......^.(.....
........
15:04:04.154558 IP localhost.35820 > localhost.19001: Flags [P.], seq 207:216, ack 192, win 350, options [nop,nop,TS val 27059371 ecr 27059369], length 9
E..=< @.@.............J9v..^.......^.1.....
..........v._...0
15:04:04.154576 IP localhost.19001 > localhost.35820: Flags [.], ack 216, win 350, options [nop,nop,TS val 27059371 ecr 27059371], length 0
E..4e.@.@...........J9......v..g...^.(.....
........
15:04:04.154754 IP localhost.35820 > localhost.19001: Flags [F.], seq 216, ack 192, win 350, options [nop,nop,TS val 27059371 ecr 27059371], length 0
E..4<!@.@.............J9v..g.......^.(.....
........
15:04:04.191775 IP localhost.19001 > localhost.35820: Flags [.], ack 217, win 350, options [nop,nop,TS val 27059381 ecr 27059371], length 0
E..4e.@.@...........J9......v..h...^.(.....
........
15:04:04.243793 IP localhost.19001 > localhost.35820: Flags [R.], seq 192, ack 217, win 350, options [nop,nop,TS val 27059394 ecr 27059371], length 0
E..4e.@.@...........J9......v..h...^.(.....
........
```

Looking at this, I'd say the problem is that if the fromNet stream closes before the protocol switch is done, then the toHttp stream is complete, but then the fromHttp stream receives the WebSocket response message and switches protocol, and then that close signal is essentially lost since the toHttp stream is now irrelevant.  One way to fix this is when fromNet finishes, both fromHttp should be cancelled.

**Comment by drewhk** [comment]
Ok, looks like a classic stage logic bug. Thanks for the reproducer!

**Comment by jroper** [comment]
One thing that isn't explained is the first pcap that I attached - no FIN was sent by the client in that case.  So maybe there are other bugs in there.  I wasn't able to reproduce that one though.

**Comment by MichaelZinsmaier** [comment]
Hey, 

I was able to reproduce a mid-connection "RST, ACK" with the Chrome WebSocket extension
but only using the __Async__ handler. Plus if __auto-fusing__ is __off__ it will only happen from time to time (seldom).

Hope that helps!

Example:

```Scala
  val requestHandler: HttpRequest => Future[HttpResponse] = {
    case req@HttpRequest(GET, Uri.Path("/ws"), _, _, _) =>
      req.header[UpgradeToWebsocket] match {
        case Some(upgrade: UpgradeToWebsocket) => {
          Future(upgrade.handleMessagesWithSinkSource(sink, src))
        }
        case None => Future(HttpResponse(404))
      }
  }

  @Test(groups = Array("demo"))
  def manualTestUsingChromeExtension(): Unit = {
    val debugPoint_connectHere = ""
    Thread.sleep(1000)
    // the connection should be open now according to chrome
    val debugPoint_SendHere = ""

    Thread.sleep(500)

    // RST should have been send
    val debugPoint_StopCapturingHere = ""
  }
```

* start the server in debug mode
* open to "ws://localhost:8080/ws"  -> chrome shows opening
* continue to the next debug point -> chrome shows open
* send some data
* continue to the next debug point
* stop capturing

=> instead of a ACK a RST, ACK is send by the server.

**Comment by rkuhn** [comment]
@jroper I’ll look into this today. Please mention me on such blockers, I do not get notifications otherwise. One thing that is not yet clear to me: when exactly do you need a fix, and on which release branch(es)?

**Comment by jroper** [comment]
We have two options, get a fix for Play 2.5 (which will use Akka streams from Akka 2.4), or we disable the tests that are impacted by this.

**Comment by rkuhn** [comment]
Well, since this is an actual bug I don’t think disabling the tests helps much :-) Play 2.5 is not out, so what is the timeline?

**Comment by rkuhn** [comment]
Looking into this I found a possible case where things could go wrong as you indicate, @jroper, but for the life of me I can neither write a reproducer nor can I find an explanation for the first TCP dump—without the connection being half-closed this error cannot happen.

Another thing: half-closing the connection when having asked for an upgrade is not going to work reliably in any case, why would that be reasonable to expect? At least this cannot be a conformant websocket client.

**Comment by jroper** [comment]
It's a bug that only affects our Akka HTTP backend, which is still marked as experimental, so I'm not as worried about disabling the tests for that - we'll get the fix eventually.

You're correct, it's not a conformant websocket client,  it's our test client which can do a lot of non conformant things.  This particular test checks that when a connection is closed non gracefully, the resources on the server are released, which was an actual bug that we had in our Netty WebSocket implementation two years ago, and we run the same tests against both Netty and Akka HTTP backends.

**Comment by jroper** [comment]
I've had another go at reproducing the RESET with no FIN first this morning.  Unfortunately, it appears we have a true heisenbug.  When I have a breakpoint set in `onUpstreamFailure` on `TcpStages.scala:271` (I have this breakpoint configured to not suspend, rather to just execute `printStackTrace` on the exception), I can't reproduce it, in every pcap, there's a FIN packet sent first, and the exception is the one above.  However, as soon as I turn the breakpoint off, I can reproduce it about 60% of the time, here's a pcap that I just got then:

```
10:32:04.153432 IP localhost.52234 > localhost.19001: Flags [S], seq 1429253376, win 43690, options [mss 65495,sackOK,TS val 44806049 ecr 0,nop,wscale 7], length 0
E..<..@.@.:..........
J9U0...........0.........
............
10:32:04.153462 IP localhost.19001 > localhost.52234: Flags [S.], seq 428202042, ack 1429253377, win 43690, options [mss 65495,sackOK,TS val 44806049 ecr 44806049,nop,wscale 7], length 0
E..<..@.@.<.........J9.
...:U0.......0.........
............
10:32:04.153478 IP localhost.52234 > localhost.19001: Flags [.], ack 1, win 342, options [nop,nop,TS val 44806049 ecr 44806049], length 0
E..4..@.@.: .........
J9U0.....;...V.(.....
........
10:32:04.158783 IP localhost.52234 > localhost.19001: Flags [P.], seq 1:207, ack 1, win 342, options [nop,nop,TS val 44806050 ecr 44806049], length 206
E.....@.@.9Q.........
J9U0.....;...V.......
........GET /stream HTTP/1.1
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Key: So96biA42qEcWLM9vuwwQw==
Host: localhost:19001
Sec-WebSocket-Origin: http://localhost:19001
Sec-WebSocket-Version: 13


10:32:04.158803 IP localhost.19001 > localhost.52234: Flags [.], ack 207, win 350, options [nop,nop,TS val 44806050 ecr 44806050], length 0
E..4w.@.@..'........J9.
...;U0.....^.(.....
........
10:32:04.262485 IP localhost.19001 > localhost.52234: Flags [P.], seq 1:192, ack 207, win 350, options [nop,nop,TS val 44806076 ecr 44806050], length 191
E...w.@.@..g........J9.
...;U0.....^.......
........HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Sec-WebSocket-Accept: d4/mU8FNm1V3f/mCkuRWRpF+4gE=
Server: akka-http/2.4.1
Date: Mon, 18 Jan 2016 23:32:04 GMT
Connection: upgrade


10:32:04.262496 IP localhost.52234 > localhost.19001: Flags [.], ack 192, win 350, options [nop,nop,TS val 44806076 ecr 44806076], length 0
E..4..@.@.:..........
J9U0.........^.(.....
........
10:32:04.269994 IP localhost.19001 > localhost.52234: Flags [R.], seq 192, ack 207, win 350, options [nop,nop,TS val 44806078 ecr 44806076], length 0
E..4w.@.@..%........J9.
....U0.....^.(.....
........
```

Any ideas for finding out why this is happening?

**Comment by jroper** [comment]
Actually, my guess is that the FIN is being sent, but is never reaching tcpdump because the RESET is causing the pcap buffer to be dropped before tcpdump is able to consume it.  RESET means forget everything you know about this connection.  When the kernel receives it, it would have to clean up all resources associated with the connection, and this would probably include the queue of packets heading to tcpdump, dropping any packets as yet unreceived by tcpdump.  One such packet is probably the FIN packet.  My understanding of libpcap is that it's asynchronous and uses buffering, so just because the application has received and handled the FIN packet, doesn't mean tcpdump has yet received it.

This explains why adding the breakpoint means you see the FIN packet in tcpdump, even though the breakpoint isn't suspending the thread, it is doing some blocking IO when it prints out the stack trace, and this probably adds enough of a delay to give tcpdump enough time to consume the FIN packet.

**Comment by jroper** [comment]
By the way, for now I've disabled these specific tests in Play, this should allow us to get a milestone of Play out with Akka streams 2, which will help certain other projects that need a Play 2.5 release too.  This takes the pressure off getting a streams 2 release out.  It's only the Akka HTTP websocket tests in Play that are affected, so I think that's fine for now.

**Comment by MichaelZinsmaier** [comment]
Hey guys,

we narrowed down our issue (a bit late) and I would like to test with the fix from the next nightly build but http://repo.akka.io/snapshots/com/typesafe/akka/ is not hosting akka-parsing-experimental_2.11:2.4-SNAPSHOT (a http-core dependency)

I wrote a small app to [reproduce](https://github.com/MichaelZinsmaier/websocket-bug) our issues. For me it looks similar to the described problem. From the network [dump](https://github.com/MichaelZinsmaier/websocket-bug/blob/master/results/ErrorExample.png) I think our small Netty client does nothing special, it should be a conformant client but every 3rd run of a echo client is unsuccessful.

It would be nice if you could either try it out or add akka-parsing to the nightly build repo, then I try it tomorrow.

Best Michael

**Comment by rkuhn** [comment]
I don’t think the parsing module is needed, have you tried excluding the dependency?

**Comment by rkuhn** [comment]
see #19517

**Comment by MichaelZinsmaier** [comment]
It compiles but does not run if the parsing module is excluded (classloader exception) however I managed to exclude the SNAPSHOT version and include the 2.0.2 release of the parsing module => that's fine

strangely our tests / the example project are working now BUT the nightly builds currently serve a version from 18.01.2016 => before the fix has been merged, it seems that our problem has been solved but not by the 19488 fix

I'll write a update tomorrow, testing with the new nightly build version (including 19488)

**Comment by rkuhn** [comment]
Thanks for trying out all the things! (including verifying that now parsing.jar is actually needed, I thought it was not)

**Comment by MichaelZinsmaier** [comment]
you are welcome

tests with the nightly build from 19.01.2016 are passing => our problem seems to be fixed
we will stay with 2.0.1 until the next release come out

thanks for your help

**Comment by jroper** [comment]
I can confirm this is now fixed in Akka 2.4.2-RC1, and have reenabled these tests in our build. Thanks!
